File: frontend/src/lib/timesheets.ts
const BASE = `${process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"}/api/v1/timesheets`;

async function req<T>(path: string, opts?: RequestInit): Promise<T> {
  const r = await fetch(`${BASE}${path}`, {
    headers: { "Content-Type": "application/json" },
    ...opts,
  });
  if (!r.ok) {
    const err = await r.json().catch(() => ({}));
    throw new Error((err as { detail?: string }).detail ?? r.statusText);
  }
  if (r.status === 204) return undefined as T;
  return r.json();
}

// ── Types ─────────────────────────────────────────────────────────────────────

export type TimesheetStatus = "draft" | "submitted" | "manager_approved" | "rejected" | "finalized";
export type ActivityType = "regular" | "overtime" | "training" | "leave" | "meeting" | "admin" | "project" | "support" | "other";
export type ApprovalAction = "submitted" | "approved" | "rejected" | "finalized" | "resubmitted";
export type TSAIAgentType = "utilization_analyzer" | "anomaly_detector";
export type TSAIRecStatus = "pending" | "acknowledged" | "actioned" | "dismissed";

export interface TimesheetLine {
  timesheet_line_id: string;
  timesheet_id: string;
  work_date: string;
  project_id: string | null;
  project_name: string | null;
  task_id: string | null;
  task_name: string | null;
  activity_type: ActivityType;
  cost_center_id: string | null;
  cost_center_name: string | null;
  hours_worked: number;
  overtime_flag: boolean;
  billable_flag: boolean;
  notes: string | null;
  created_at: string;
}

export interface ApprovalLog {
  approval_id: string;
  timesheet_id: string;
  approver_id: string | null;
  approver_name: string | null;
  action: ApprovalAction;
  action_date: string;
  comments: string | null;
}

export interface TimesheetHeader {
  timesheet_id: string;
  employee_id: string;
  employee_name: string | null;
  department_id: string | null;
  department_name: string | null;
  manager_id: string | null;
  manager_name: string | null;
  period_start: string;
  period_end: string;
  status: TimesheetStatus;
  total_hours: number;
  overtime_hours: number;
  regular_hours: number;
  submitted_at: string | null;
  approved_by: string | null;
  approved_at: string | null;
  finalized_at: string | null;
  rejection_count: number;
  notes: string | null;
  created_at: string;
  lines: TimesheetLine[];
  approval_logs: ApprovalLog[];
}

export interface TimesheetDashboard {
  total_timesheets: number;
  draft: number;
  submitted: number;
  manager_approved: number;
  rejected: number;
  finalized: number;
  total_hours_this_period: number;
  overtime_hours_this_period: number;
  avg_hours_per_employee: number;
  pending_approval: number;
  employees_with_timesheets: number;
  project_hours: { project: string; hours: number }[];
}

export interface TSAIRec {
  rec_id: string;
  agent_type: TSAIAgentType;
  status: TSAIRecStatus;
  title: string;
  body: string;
  employee_id: string | null;
  department_id: string | null;
  timesheet_id: string | null;
  score: number | null;
  actioned_by: string | null;
  actioned_at: string | null;
  created_at: string;
}

// ── Maps ──────────────────────────────────────────────────────────────────────

export const STATUS_COLOR: Record<TimesheetStatus, string> = {
  draft: "bg-gray-100 text-gray-700",
  submitted: "bg-blue-100 text-blue-700",
  manager_approved: "bg-purple-100 text-purple-700",
  rejected: "bg-red-100 text-red-700",
  finalized: "bg-green-100 text-green-700",
};

export const STATUS_LABEL: Record<TimesheetStatus, string> = {
  draft: "Draft",
  submitted: "Submitted",
  manager_approved: "Manager Approved",
  rejected: "Rejected",
  finalized: "Finalized",
};

export const ACTIVITY_COLOR: Record<ActivityType, string> = {
  regular: "bg-blue-100 text-blue-700",
  overtime: "bg-orange-100 text-orange-700",
  training: "bg-purple-100 text-purple-700",
  leave: "bg-yellow-100 text-yellow-700",
  meeting: "bg-cyan-100 text-cyan-700",
  admin: "bg-gray-100 text-gray-600",
  project: "bg-green-100 text-green-700",
  support: "bg-teal-100 text-teal-700",
  other: "bg-gray-100 text-gray-500",
};

export const ACTIVITY_TYPES: ActivityType[] = [
  "regular","overtime","training","leave","meeting","admin","project","support","other"
];

export const ACTION_COLOR: Record<ApprovalAction, string> = {
  submitted: "text-blue-600",
  approved: "text-green-600",
  rejected: "text-red-600",
  finalized: "text-purple-600",
  resubmitted: "text-yellow-600",
};

// ── API ───────────────────────────────────────────────────────────────────────

export const timesheetsApi = {
  getDashboard: () => req<TimesheetDashboard>("/dashboard"),

  create: (d: object) => req<TimesheetHeader>("/", { method: "POST", body: JSON.stringify(d) }),
  list: (params?: { employee_id?: string; status?: string; period_start?: string; department_id?: string }) => {
    const q = new URLSearchParams();
    if (params?.employee_id) q.set("employee_id", params.employee_id);
    if (params?.status) q.set("status", params.status);
    if (params?.period_start) q.set("period_start", params.period_start);
    if (params?.department_id) q.set("department_id", params.department_id);
    return req<TimesheetHeader[]>(`/?${q}`);
  },
  get: (id: string) => req<TimesheetHeader>(`/${id}`),
  submit: (id: string, d: object) => req<TimesheetHeader>(`/${id}/submit`, { method: "POST", body: JSON.stringify(d) }),
  approve: (id: string, d: object) => req<TimesheetHeader>(`/${id}/approve`, { method: "POST", body: JSON.stringify(d) }),
  reject: (id: string, d: object) => req<TimesheetHeader>(`/${id}/reject`, { method: "POST", body: JSON.stringify(d) }),
  finalize: (id: string, d: object) => req<TimesheetHeader>(`/${id}/finalize`, { method: "POST", body: JSON.stringify(d) }),

  addLine: (id: string, d: object) => req<TimesheetLine>(`/${id}/lines`, { method: "POST", body: JSON.stringify(d) }),
  updateLine: (lineId: string, d: object) => req<TimesheetLine>(`/lines/${lineId}`, { method: "PATCH", body: JSON.stringify(d) }),
  deleteLine: (lineId: string) => req<void>(`/lines/${lineId}`, { method: "DELETE" }),
  autoFill: (id: string, attendance: object[]) =>
    req<{ lines_added: number }>(`/${id}/auto-fill`, { method: "POST", body: JSON.stringify(attendance) }),

  approvalQueue: (department_id?: string) =>
    req<TimesheetHeader[]>(`/approval-queue/pending${department_id ? `?department_id=${department_id}` : ""}`),

  reportSummary: (period_start?: string) =>
    req<object>(`/reports/summary${period_start ? `?period_start=${period_start}` : ""}`),
  reportUtilization: (period_start?: string) =>
    req<object[]>(`/reports/utilization${period_start ? `?period_start=${period_start}` : ""}`),
  reportOvertime: (period_start?: string) =>
    req<object[]>(`/reports/overtime${period_start ? `?period_start=${period_start}` : ""}`),
  reportProjectTime: (project_id?: string) =>
    req<object[]>(`/reports/project-time${project_id ? `?project_id=${project_id}` : ""}`),
  reportActivitySummary: (period_start?: string) =>
    req<object[]>(`/reports/activity-summary${period_start ? `?period_start=${period_start}` : ""}`),
  payrollInput: (period_start: string, period_end: string) =>
    req<object[]>(`/payroll-input?period_start=${period_start}&period_end=${period_end}`),

  runUtilizationAnalyzer: () => req<TSAIRec[]>("/ai/run-utilization-analyzer", { method: "POST" }),
  runAnomalyDetector: () => req<TSAIRec[]>("/ai/run-anomaly-detector", { method: "POST" }),
  listAIRecs: (status?: string) => req<TSAIRec[]>(`/ai/recommendations${status ? `?status=${status}` : ""}`),
  ackAIRec: (id: string, d: object) => req<TSAIRec>(`/ai/recommendations/${id}`, { method: "PATCH", body: JSON.stringify(d) }),
};

File: backend/app/models/timesheets.py
from __future__ import annotations
import uuid
from datetime import datetime, date
from decimal import Decimal
from enum import Enum as PyEnum
from sqlalchemy import (
    Column, String, Numeric, Boolean, DateTime, Date,
    ForeignKey, Text, Integer, Enum, UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.db.base import Base


class TimesheetStatus(str, PyEnum):
    DRAFT = "draft"
    SUBMITTED = "submitted"
    MANAGER_APPROVED = "manager_approved"
    REJECTED = "rejected"
    FINALIZED = "finalized"


class ActivityType(str, PyEnum):
    REGULAR = "regular"
    OVERTIME = "overtime"
    TRAINING = "training"
    LEAVE = "leave"
    MEETING = "meeting"
    ADMIN = "admin"
    PROJECT = "project"
    SUPPORT = "support"
    OTHER = "other"


class ApprovalAction(str, PyEnum):
    SUBMITTED = "submitted"
    APPROVED = "approved"
    REJECTED = "rejected"
    FINALIZED = "finalized"
    RESUBMITTED = "resubmitted"


class TSAIAgentType(str, PyEnum):
    UTILIZATION_ANALYZER = "utilization_analyzer"
    ANOMALY_DETECTOR = "anomaly_detector"


class TSAIRecStatus(str, PyEnum):
    PENDING = "pending"
    ACKNOWLEDGED = "acknowledged"
    ACTIONED = "actioned"
    DISMISSED = "dismissed"


class TimesheetHeader(Base):
    __tablename__ = "timesheet_headers"

    timesheet_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    employee_id = Column(String(100), nullable=False)
    employee_name = Column(String(200), nullable=True)
    department_id = Column(String(100), nullable=True)
    department_name = Column(String(200), nullable=True)
    manager_id = Column(String(100), nullable=True)
    manager_name = Column(String(200), nullable=True)
    period_start = Column(Date, nullable=False)
    period_end = Column(Date, nullable=False)
    status = Column(Enum(TimesheetStatus), default=TimesheetStatus.DRAFT, nullable=False)
    total_hours = Column(Numeric(6, 2), default=Decimal("0.00"))
    overtime_hours = Column(Numeric(6, 2), default=Decimal("0.00"))
    regular_hours = Column(Numeric(6, 2), default=Decimal("0.00"))
    submitted_at = Column(DateTime, nullable=True)
    approved_by = Column(String(200), nullable=True)
    approved_at = Column(DateTime, nullable=True)
    finalized_at = Column(DateTime, nullable=True)
    rejection_count = Column(Integer, default=0)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    lines = relationship("TimesheetLine", back_populates="timesheet", cascade="all, delete-orphan")
    approval_logs = relationship("TimesheetApprovalLog", back_populates="timesheet", cascade="all, delete-orphan")

    __table_args__ = (
        UniqueConstraint("employee_id", "period_start", "period_end", name="uq_timesheet_employee_period"),
    )


class TimesheetLine(Base):
    __tablename__ = "timesheet_lines"

    timesheet_line_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    timesheet_id = Column(UUID(as_uuid=True), ForeignKey("timesheet_headers.timesheet_id"), nullable=False)
    work_date = Column(Date, nullable=False)
    project_id = Column(String(100), nullable=True)
    project_name = Column(String(200), nullable=True)
    task_id = Column(String(100), nullable=True)
    task_name = Column(String(200), nullable=True)
    activity_type = Column(Enum(ActivityType), default=ActivityType.REGULAR, nullable=False)
    cost_center_id = Column(String(100), nullable=True)
    cost_center_name = Column(String(200), nullable=True)
    hours_worked = Column(Numeric(5, 2), nullable=False)
    overtime_flag = Column(Boolean, default=False)
    billable_flag = Column(Boolean, default=False)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)

    timesheet = relationship("TimesheetHeader", back_populates="lines")


class TimesheetApprovalLog(Base):
    __tablename__ = "timesheet_approval_logs"

    approval_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    timesheet_id = Column(UUID(as_uuid=True), ForeignKey("timesheet_headers.timesheet_id"), nullable=False)
    approver_id = Column(String(100), nullable=True)
    approver_name = Column(String(200), nullable=True)
    action = Column(Enum(ApprovalAction), nullable=False)
    action_date = Column(DateTime, default=datetime.utcnow)
    comments = Column(Text, nullable=True)

    timesheet = relationship("TimesheetHeader", back_populates="approval_logs")


class TSAIRecommendation(Base):
    __tablename__ = "ts_ai_recommendations"

    rec_id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    agent_type = Column(Enum(TSAIAgentType), nullable=False)
    status = Column(Enum(TSAIRecStatus), default=TSAIRecStatus.PENDING, nullable=False)
    title = Column(String(300), nullable=False)
    body = Column(Text, nullable=False)
    employee_id = Column(String(100), nullable=True)
    department_id = Column(String(100), nullable=True)
    timesheet_id = Column(UUID(as_uuid=True), nullable=True)
    score = Column(Integer, nullable=True)
    actioned_by = Column(String(200), nullable=True)
    actioned_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
